### PR TITLE
Use editable customer and avoid loading unnecessary data

### LINKED ISSUE #25683: Customers Query slow with many orders
#### Describe the bug

Hello there,

When trying to view or edit the password of a customer with many orders, this generates slowness in seeing customer as edit (Change Password).

You should not delay in the edit client option (Change Password) since in this option there is only basic information, there is no detailed information about the client.

#### Expected behavior

Explain what you expected to happen instead.

#### Steps to Reproduce

1. Go to 'Clients'
2. Click on 'Clients'
3. Find a customer who has thousands of orders

**Screenshots**

image

https://drive.google.com/file/d/1-yz5mDkFm9_yc-In4JJaxuVBc9qUsXOV/view?usp=sharing

#### Additional information

* PrestaShop version: 1.7.7.7
* PHP version: 7.3

### PR DESCRIPTION
**Pull request by Hlavtox**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Prestashop was loading `ViewableCustomer` with a lot of unnecessary information (hundreds of queries and slowing down the opening of the edit form) instead of lightweight `EditableCustomer`. I also needed to add `isGuest` to `EditableCustomer` due to it being used in the template, we will need it anyway for other things.
| Type?             | improvement
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes part of #25683
| How to test?      | Try to edit some customers and see its working fine.
| Possible impacts? | -

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/25686)

### CODE REVIEW

File: src/Core/Domain/Customer/QueryResult/EditableCustomer.php [hunk 1]
```diff
@@ -155,6 +161,7 @@ public function __construct(
         Email $email,
         Birthday $birthday,
         $isEnabled,
+        $isGuest,
```
Comment (Progi1984) [inline src/Core/Domain/Customer/QueryResult/EditableCustomer.php]: Could you put this parameter at the end ?

Comment (Progi1984) [reply src/Core/Domain/Customer/QueryResult/EditableCustomer.php]:
```suggestion
        bool $isGuest,
```

File: src/Core/Domain/Customer/QueryResult/EditableCustomer.php [hunk 2]
```diff
@@ -243,6 +251,14 @@ public function isEnabled()
         return $this->isEnabled;
     }
 
+    /**
+     * @return bool
+     */
+    public function isGuest()
```
Comment (Progi1984) [inline src/Core/Domain/Customer/QueryResult/EditableCustomer.php]:
```suggestion
    public function isGuest(): bool
```

**Review by Progi1984 — COMMENTED** [review]
@Hlavtox I confirm it's a BC Break as you add a new parameter to a constructor without default value.

**Comment by Hlavtox** [comment]
@Progi1984 Updated and tests fixed. I also added the guest property to CustomerFormDataProvider, so it's ready to be used in the form.

**Comment by matthieu-rolland** [comment]
> @Progi1984 Updated and tests fixed. I also added the guest property to CustomerFormDataProvider, so it's ready to be used in the form.

Unfortunately, it's still a breaking change since the signature of the class `PrestaShop\PrestaShop\Core\Domain\Customer\QueryResult\EditableCustomer` is changing

Could you declare this BC break in the PR, below the description table? Sometimes BC breaks are approved if it's really necessary or very valuable, which could be the case here

**EDIT**:  just noticed you filled it on the table 👍

**Review by sowbiba — DISMISSED** [review]
LGTM

**Review by Progi1984 — DISMISSED** [review]

File: src/Core/Domain/Customer/QueryResult/EditableCustomer.php [hunk 3]
```diff
@@ -165,7 +171,8 @@ public function __construct(
         $website,
         $allowedOutstandingAmount,
         $maxPaymentDays,
-        $riskId
+        $riskId,
+        bool $isGuest
```
Comment (PierreRambaud) [inline src/Core/Domain/Customer/QueryResult/EditableCustomer.php]: To prevent bc breaks, you can maybe set this parameter as false by default?

Comment (Hlavtox) [reply src/Core/Domain/Customer/QueryResult/EditableCustomer.php]: @PierreRambaud Whatever you think is best. Should I do it? @Progi1984

Comment (PierreRambaud) [reply src/Core/Domain/Customer/QueryResult/EditableCustomer.php]: That's also the value you pass by default in form

Comment (Hlavtox) [reply src/Core/Domain/Customer/QueryResult/EditableCustomer.php]: @PierreRambaud Updated

File: src/PrestaShopBundle/Controller/Admin/Sell/Customer/CustomerController.php [hunk 4]
```diff
@@ -208,8 +208,8 @@ public function createAction(Request $request)
     public function editAction($customerId, Request $request)
     {
         $this->addGroupSelectionToRequest($request);
-        /** @var ViewableCustomer $customerInformation */
-        $customerInformation = $this->getQueryBus()->handle(new GetCustomerForViewing((int) $customerId));
+        /** @var EditableCustomer $customerInformation */
+        $customerInformation = $this->getQueryBus()->handle(new GetCustomerForEditing((int) $customerId));
```
Comment (PierreRambaud) [inline src/PrestaShopBundle/Controller/Admin/Sell/Customer/CustomerController.php:212]: Just need confirmation from @jolelievre / @matks about this change.

**Review by Progi1984 — DISMISSED** [review]

File: src/PrestaShopBundle/Controller/Admin/Sell/Customer/CustomerController.php (hunk 4 above)
Comment (matks) [reply src/PrestaShopBundle/Controller/Admin/Sell/Customer/CustomerController.php:212]: The change is relevant IMO.

**Review by matks — DISMISSED** [review]

**Review by PierreRambaud — DISMISSED** [review]

**Comment by hibatallahAouadni** [comment]
This PR need the Dev Team's confirmation on the issue :pray:

**Comment by matks** [comment]
> This PR need the Dev Team's confirmation on the issue 🙏

I validate the issue.

**Comment by hibatallahAouadni** [comment]
Hello @Hlavtox 

Thanks for your PR!
### Without PR:
I use **one Customer** with **2 Adresses**, **6 cart rules** and **13 orders** => **Error 500**

https://watch.screencastify.com/v/8KgPI6ba2VUsVYmLFDb0

### With PR:
I use **one Customer** with **2 Adresses**, **6 cart rules** and **22 order**s => it tooks 3 secondes to load the page

https://watch.screencastify.com/v/dnebNlmduSQMNjFZfh0W

Ping @PrestaShop/qa-team wdyt? 

Thanks!

**Comment by hibatallahAouadni** [comment]
Ping @Hlavtox after checking with @khouloudbelguith we have the same results :tada: 
You just need resolve the Conflicting files :wink: 

Thanks!

**Review by jolelievre — COMMENTED** [review]
Hi @Hlavtox,

thanks for your contribution, you're right it's weird the the view command was used for the edit form. All your modifications make sense, however could you complete you PR with updated behat tests as well please?

- first the assertion test should validate the new `isGuest` field https://github.com/PrestaShop/PrestaShop/blob/3d06748975e00c548e940bf82862d9d211ae94d4/tests/Integration/Behaviour/Features/Context/CustomerManagerFeatureContext.php#L213
- second, in order to test this new field correctly you need to make it change
- this is a good opportunity to test the `TransformGuestToCustomerCommand` command to change this value
- https://github.com/PrestaShop/PrestaShop/blob/3d06748975e00c548e940bf82862d9d211ae94d4/tests/Integration/Behaviour/Features/Context/CustomerManagerFeatureContext.php#L175 the behat statement already exists, but surprisingly it is never used
- the only thing missing, I didn't check, may be to create a guest
- seems like you can create a guest customer by using the Guest group instead of the Customer group via the `AddCustomerCommand` so you might just rely on this for now
- Ideally we should have two different commands AddCustomerCommand and AddGuestCommand that would reflect the business logic, but it might be too much for this PR, it could be a good refacto in the future though

I realize it's a bit more work, but we should try and keep the tests on CQRS commands as detailed as possible ^^

**Comment by jolelievre** [comment]
By the way @Hlavtox sorry for the late review, I just saw that the PR is almost a month old but I didn't see it before. I still think the suggestions I made would improve its quality a lot (that's why I didn't "block it" with expected changes). Sorry for the trouble 😅

File: templates/bundles/PrestaShopBundle/Admin/Sell/Customer/edit.html.twig [hunk 5]
```diff
@@ -24,15 +24,15 @@
...
+{% set fullName = '%s. %s'|format(customerInformation.firstName.getValue[:1], customerInformation.lastName.getValue) %}
 {% set layoutTitle = 'Editing customer %name%'|trans({'%name%': fullName}, 'Admin.Orderscustomers.Feature') %}
 
 {% extends '@PrestaShop/Admin/layout.html.twig' %}
 
 {% block content %}
   <div class="row justify-content-center">
     <div class="col">
-      {% include '@PrestaShop/Admin/Sell/Customer/Blocks/form.html.twig' with {'isGuest': customerInformation.personalInformation.guest} %}
+      {% include '@PrestaShop/Admin/Sell/Customer/Blocks/form.html.twig' with {'isGuest': customerInformation.isGuest} %}
```
Comment (jolelievre) [inline templates/bundles/PrestaShopBundle/Admin/Sell/Customer/edit.html.twig]: Side note, I just checked in the form seems like this `isGuest` variable has no use 🤔

File: templates/bundles/PrestaShopBundle/Admin/Sell/Customer/edit.html.twig [hunk 6]
```diff
@@ -24,15 +24,15 @@
  *#}
 
 {% set enableSidebar = true %}
-{% set fullName = '%s. %s'|format(customerInformation.personalInformation.firstName[:1], customerInformation.personalInformation.lastName) %}
+{% set fullName = '%s. %s'|format(customerInformation.firstName.getValue[:1], customerInformation.lastName.getValue) %}
```
Comment (jolelievre) [inline templates/bundles/PrestaShopBundle/Admin/Sell/Customer/edit.html.twig]: Oh, I didn't know this notation to get the first letter, good to know!!

**Comment by jolelievre** [comment]
This also made me realize that we have a (seemingly) useless `Customer::id_guest` field that is only used here https://github.com/PrestaShop/PrestaShop/blob/96a17720ff2e0b9bf557e2febb2dd77e57406171/config/config.inc.php#L256

Maybe a candidate for a deprecation/removal (not the scope of this PR of course)

**Comment by Hlavtox** [comment]
Hi @jolelievre, thanks for checking the PR! :-) I will reply in points.

**isGuest**
This variable is currently only used in the template now, but we will need it to be included in **EditableCustomer**, so we can actually **create guests from backoffice**, which is not possible now.

**is_guest** is basically the most important thing that differentiates guests from customers. Them being in different groups is not that important.

I already have a concept of guest creation locally (working 👍), I will open a new issue with this and make a specs for it, then we can decide if we make a separate command for guest creation or just make some conditions in the current command.

Výstřižek

After this PR, everything will be prepared for this guest creation, you will just need to add the field to the form. :-)

I have zero skills with the testing, so, do we do them now or we agree to add tests in bigger, guest creation PR? Can I ask @boubkerbribri for help?

**Comment by boubkerbribri** [comment]
Hello @Hlavtox, 
> I have zero skills with the testing, so, do we do them now or we agree to add tests in bigger, guest creation PR? Can I ask @boubkerbribri for help?

Yes, of course, when everything is ready, I will need @PrestaShop/qa-functional to write new scenarios for our tests campaign, then we can implement them.

**Comment by jolelievre** [comment]
Hi @Hlavtox 
the tests I was mentioning were behat tests for the CQRS commands and queries I wasn't taking about E2E tests that can be added later indeed

I'd prefer to integrate the behat tests in this PR as it will enforce and validate the new modifications, and I'm afraid we will forget about it if it's not added now But if you're not comfortable with behat test I'll be happy to add them to the PR if you want?

**Comment by Hlavtox** [comment]
@jolelievre Whatever you think is the best for this! Please, go ahead and do it, so we can merge it as soon as possible. ;-) Thank you! 🍺

**Comment by jolelievre** [comment]
Hi @Hlavtox,

I improved the behat tests for customer commands, most of the fields were not tested. However I couldn't test the `is_guest` much more because there is no way to create a guest or to edit the `is_guest` field for now, and I didn't want to increase the initial scope of this PR.

So more tests will be needed later I guess when guests will be handled, I'm still not sure if the `is_guest` should be editable. I get that turning a guest into a customer is important but I'm not sure the other way around is a real use case. If so then we should add a dedicated `AddGuestCommand` to make sure the guest is properly created with the right values and groups.

**Comment by Hlavtox** [comment]
@jolelievre Thank you man! Let's discuss it in the new PR, I will open immediately after merging this. :-)

@hibatallahAouadni @khouloudbelguith Ladies you need to test customer edit to see the benefit of the PR -> click the pencil icon in the customer list or from customer detail. Without this PR, it will throw 500 or take X seconds. With this PR, it will be absolutely instant. Check again please, before we merge. :-)

**Review by jolelievre — APPROVED** [review]

**Review by kpodemski — APPROVED** [review]

**Comment by hibatallahAouadni** [comment]
Hello @Hlavtox & @jolelievre 

Thanks guys for your hard working :clap: 
I use one Customer with 2 Adresses, 6 cart rules and 22 orders:

- **With PR**, the edit Customer page displays right away and it tooks less than 3 secondes to load the page View Customer :tada: 
- **Without PR**, it tooks more than 3 secondes to display the edit Customer, the View Customer page displays 500 error page :disappointed: 

QA :heavy_check_mark: 
Thanks!

**Comment by kpodemski** [comment]
Thank you everyone involved 👍

**Comment by Hlavtox** [comment]
Thanks everyone!

**Pull request merged**